clash-prelude/src/Clash/Sized/Vector.hs

```diff
@@ -1058,8 +1058,26 @@ foldr f z (x `Cons` xs) = f x (foldr f z xs)
 -- delay, of O(@'length' xs@). Use 'fold' if your binary operator @f@ is
 -- associative, as @"'fold' f xs"@ produces a structure with a depth of
 -- O(log_2(@'length' xs@)).
-foldl :: (b -> a -> b) -> b -> Vec n a -> b
-foldl f z xs = last (scanl f z xs)
+foldl :: forall b a n . (b -> a -> b) -> b -> Vec n a -> b
+foldl f z0 xs0
+  -- We use 'go' to make 'foldl' strict during Haskell simulation, while using
+  -- 'scanl' to make Clash's life easier during normalization. We make 'foldl'
+  -- strict due to similar reasoning to:
+  --
+  --   https://github.com/hasura/graphql-engine/pull/2933#discussion_r328821960
+  --
+  -- Also see:
+  --
+  --   https://well-typed.com/blog/90/
```

Comment:
But it was posted on April 1st....

Anyhow, I feel this one is the stronger reference to make the change, as it actually argues to make a strict foldl the default. i.e. I would reference it first.